Mathlib/AlgebraicGeometry/EllipticCurve/NormalForms.lean

```diff
@@ -0,0 +1,658 @@
+  rw [self.b₆]
+  linear_combination 2 * W.a₆ * CharP.cast_eq_zero R 2
+
+theorem b₈_of_char_two : W.b₈ = W.a₆ := by
+  rw [self.b₈]
+  linear_combination 2 * W.a₂ * W.a₆ * CharP.cast_eq_zero R 2
+
+theorem c₄_of_char_two : W.c₄ = 1 := by
+  rw [self.c₄, self.b₂_of_char_two]
+  ring1
+
+theorem c₆_of_char_two : W.c₆ = 1 := by
+  rw [self.c₆, self.b₂_of_char_two]
+  linear_combination (-1 - 432 * W.a₆) * CharP.cast_eq_zero R 2
+
+theorem Δ_of_char_two : W.Δ = W.a₆ := by
+  rw [WeierstrassCurve.Δ, self.b₂_of_char_two, self.b₄, self.b₆_of_char_two, self.b₈_of_char_two]
+  linear_combination -W.a₆ * CharP.cast_eq_zero R 2
+
+end IsCharTwoJNeZeroNF
+
+namespace IsCharTwoJNeZeroNF
+
+variable {E} (self : E.IsCharTwoJNeZeroNF)
+include self
+
+theorem j_of_char_two [CharP F 2] : E.j = 1 / E.a₆ := by
+  rw [EllipticCurve.j, Units.val_inv_eq_inv_val, ← div_eq_inv_mul, E.coe_Δ',
+    self.c₄_of_char_two, self.Δ_of_char_two, one_pow]
+
+theorem j_ne_zero_of_char_two [CharP F 2] : E.j ≠ 0 := by
+  rw [self.j_of_char_two, div_ne_zero_iff]
+  have h := E.Δ'.ne_zero
+  rw [E.coe_Δ', self.Δ_of_char_two] at h
+  exact ⟨one_ne_zero, h⟩
+
+end IsCharTwoJNeZeroNF
+
+/-! ### Normal forms of characteristic = 2 and j = 0 -/
+
+/-- A `WeierstrassCurve` is in normal form of characteristic = 2 and j = 0, if its $a_1, a_2 = 0$.
+In other words it is $Y^2 + a_3Y = X^3 + a_4X + a_6$. -/
+@[mk_iff]
+structure IsCharTwoJZeroNF : Prop where
+  a₁ : W.a₁ = 0
+  a₂ : W.a₂ = 0
+
+namespace IsCharTwoJZeroNF
+
+variable {W} (self : W.IsCharTwoJZeroNF)
+include self
+
+theorem b₂ : W.b₂ = 0 := by
+  rw [WeierstrassCurve.b₂, self.a₁, self.a₂]
+  ring1
+
+theorem b₄ : W.b₄ = 2 * W.a₄ := by
+  rw [WeierstrassCurve.b₄, self.a₁]
+  ring1
+
+theorem b₈ : W.b₈ = -W.a₄ ^ 2 := by
+  rw [WeierstrassCurve.b₈, self.a₁, self.a₂]
+  ring1
+
+theorem c₄ : W.c₄ = -48 * W.a₄ := by
+  rw [WeierstrassCurve.c₄, self.b₂, self.b₄]
+  ring1
+
+theorem c₆ : W.c₆ = -216 * W.b₆ := by
+  rw [WeierstrassCurve.c₆, self.b₂, self.b₄]
+  ring1
+
+theorem Δ : W.Δ = -64 * W.a₄ ^ 3 - 27 * W.b₆ ^ 2 := by
+  rw [WeierstrassCurve.Δ, self.b₂, self.b₄]
+  ring1
+
+variable [CharP R 2]
+
+theorem b₄_of_char_two : W.b₄ = 0 := by
+  rw [self.b₄]
+  linear_combination W.a₄ * CharP.cast_eq_zero R 2
+
+theorem b₈_of_char_two : W.b₈ = W.a₄ ^ 2 := by
+  rw [self.b₈]
+  linear_combination -W.a₄ ^ 2 * CharP.cast_eq_zero R 2
+
+theorem c₄_of_char_two : W.c₄ = 0 := by
+  rw [self.c₄]
+  linear_combination -24 * W.a₄ * CharP.cast_eq_zero R 2
+
+theorem c₆_of_char_two : W.c₆ = 0 := by
+  rw [self.c₆]
+  linear_combination -108 * W.b₆ * CharP.cast_eq_zero R 2
+
+theorem Δ_of_char_two : W.Δ = W.a₃ ^ 4 := by
+  rw [self.Δ, W.b₆_of_char_two]
+  linear_combination (-32 * W.a₄ ^ 3 - 14 * W.a₃ ^ 4) * CharP.cast_eq_zero R 2
+
+end IsCharTwoJZeroNF
+
+namespace IsCharTwoJZeroNF
+
+variable {E} (self : E.IsCharTwoJZeroNF)
+include self
+
+theorem j_of_char_two [CharP F 2] : E.j = 0 := by
+  simp [EllipticCurve.j, self.c₄_of_char_two]
```

Comment:
Do you want this too?
```
theorem j : E.j = -110592 * E.a₄ ^ 3 / (-64 * E.a₄ ^ 3 - 27 * E.b₆ ^ 2) := by
  have h := E.Δ'.ne_zero
  rw [E.coe_Δ', self.Δ] at h
  rw [EllipticCurve.j, Units.val_inv_eq_inv_val, ← div_eq_inv_mul, E.coe_Δ',
    self.c₄, self.Δ, div_eq_div_iff h h]
  ring1
```
I just copied this from `IsCharThreeJNeZeroNF`, but you can probably make the proof nicer.

Reply:
Nice observation! I think this is similar to that of `IsShortNF`.